File: Winter2022/labs/java/Lab6-Threading/ThreadingLabSkeleton/app/src/main/java/mobiledev/unb/ca/threadinglab/DownloaderTask.java
package mobiledev.unb.ca.threadinglab;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;

import mobiledev.unb.ca.threadinglab.model.GeoData;
import mobiledev.unb.ca.threadinglab.util.JsonUtils;

public class DownloaderTask {
    private static final int DOWNLOAD_TIME = 4;      // Download time simulation

    private final Context appContext;

    private final GeoDataListActivity activity;
    private Button refreshButton;
    private ProgressBar progressBar;
    private RecyclerView recyclerView;

    public DownloaderTask(GeoDataListActivity activity) {
        this.activity = activity;
        appContext = activity.getApplicationContext();
    }

    public DownloaderTask setRefreshButton(Button refreshButton) {
        this.refreshButton = refreshButton;
        return this;
    }

    public DownloaderTask setProgressBar(ProgressBar progressBar) {
        this.progressBar = progressBar;
        return this;
    }

    public DownloaderTask setupRecyclerView(RecyclerView recyclerView) {
        this.recyclerView = recyclerView;
        return this;
    }

    public void execute() {
        // Disable the button
        refreshButton.setEnabled(false);

        // Show the progress bar
        progressBar.setIndeterminate(false);
        progressBar.setProgress(0);
        progressBar.setMax(DOWNLOAD_TIME);
        progressBar.setVisibility(View.VISIBLE);

        // Perform background call to read the information from the URL
        Executors.newSingleThreadExecutor().execute(() -> {
            Handler mainHandler = new Handler(Looper.getMainLooper());

            // Perform the background work
            JsonUtils jsonUtils = new JsonUtils();
            ArrayList<GeoData> mGeoDataList = jsonUtils.getGeoData();
            if (null == mGeoDataList) {
                Toast.makeText(appContext, activity.getString(R.string.download_error_msg), Toast.LENGTH_SHORT).show();
            }

            // Simulating long-running operation
            for (int i = 1; i < DOWNLOAD_TIME; i++) {
                sleep();
                final int step = i;
                mainHandler.post(() -> progressBar.setProgress(step * 10));
            }

            // Update the UI with the results
            mainHandler.post(() -> setUpListView(mGeoDataList));
        });
    }

    private void sleep() {
        try {
            int mDelay = 500;
            Thread.sleep(mDelay);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    private void setUpListView(ArrayList<GeoData> mGeoDataList) {
        //  Now that the download is complete, enable the button again
        refreshButton.setEnabled(true);

        // TODO
        //  Reset the progress bar, and make it disappear
        progressBar.setProgress(0);
        progressBar.setVisibility(View.INVISIBLE);

        // TODO
        //  Setup the RecyclerView
        setupRecyclerView(mGeoDataList);

        // TODO
        //  Create a Toast indicating that the download is complete. Set its text
        //  to be the result String from doInBackground
        Toast.makeText(appContext, activity.getString(R.string.download_complete), Toast.LENGTH_SHORT).show();
    }

    private void setupRecyclerView(List<GeoData> mGeoDataList) {
        recyclerView.setAdapter(new SimpleItemRecyclerViewAdapter(mGeoDataList, activity, activity.isTwoPane()));
    }
}

File: Fall2022/lab-solutions/Lab6/ThreadingLabSkeleton/app/src/main/java/mobiledev/unb/ca/threadinglab/DownloaderTask.kt
package mobiledev.unb.ca.threadinglab

import android.content.Context
import android.os.Handler
import android.widget.ProgressBar
import androidx.recyclerview.widget.RecyclerView
import android.os.Looper
import android.widget.Button
import android.widget.Toast
import mobiledev.unb.ca.threadinglab.model.GeoData
import mobiledev.unb.ca.threadinglab.util.JsonUtils
import java.util.ArrayList
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class DownloaderTask(private val activity: GeoDataListActivity) {
    private val appContext: Context = activity.applicationContext
    private var refreshButton: Button? = null
    private var progressBar: ProgressBar? = null
    private var recyclerView: RecyclerView? = null

    fun setRefreshButton(refreshButton: Button?): DownloaderTask {
        this.refreshButton = refreshButton
        return this
    }

    fun setProgressBar(progressBar: ProgressBar?): DownloaderTask {
        this.progressBar = progressBar
        return this
    }

    fun setRecyclerView(recyclerView: RecyclerView?): DownloaderTask {
        this.recyclerView = recyclerView
        return this
    }

    fun execute() {
        // TODO
        //  Disable the button so it can't be clicked again once a download has been started
        //  Hint: Button is subclass of TextView. Read this document to see how to disable it.
        //  http://developer.android.com/reference/android/widget/TextView.html
        refreshButton!!.isEnabled = false

        // TODO
        //  Set the progress bar's maximum to be DOWNLOAD_TIME, its initial progress to be
        //  0, and also make sure it's visible.
        //  Hint: Read the documentation on ProgressBar
        //  http://developer.android.com/reference/android/widget/ProgressBar.html
        progressBar!!.isIndeterminate = false
        progressBar!!.progress = 0
        progressBar!!.max = DOWNLOAD_TIME
        progressBar!!.visibility = ProgressBar.VISIBLE

        // Perform background call to read the information from the URL
        Executors.newSingleThreadExecutor().execute {
            val handler = Handler(Looper.getMainLooper())

            // TODO
            //  Create an instance of JsonUtils and get the data from it,
            //  store the data in mGeoDataList
            val jsonUtils = JsonUtils()
            val mGeoDataList: ArrayList<GeoData>? = jsonUtils.getGeoData()
            if (null == mGeoDataList) {
                Toast.makeText(appContext,
                    activity.getString(R.string.download_error_msg),
                    Toast.LENGTH_SHORT).show()
            }

            // Simulating long-running operation
            for (i in 1..DOWNLOAD_TIME) {
                sleep()
                // TODO
                //  Update the progress bar using values
                handler.post { progressBar!!.progress = i * 10 }
            }

            // TODO
            //  Using the updateDisplay method update the UI with the results
            handler.post { updateDisplay(mGeoDataList!!) }
        }
    }

    private fun sleep() {
        try {
            val mDelay = 500
            TimeUnit.MILLISECONDS.sleep(mDelay.toLong())
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    private fun updateDisplay(mGeoDataList: ArrayList<GeoData>) {
        // TODO
        //  With the download completed, enable the button again
        refreshButton!!.isEnabled = true

        // TODO
        //  Reset the progress bar, and make it disappear
        progressBar!!.progress = 0
        progressBar!!.visibility = ProgressBar.INVISIBLE

        // TODO
        //  Setup the RecyclerView
        setupRecyclerView(mGeoDataList)

        // TODO
        //  Create a Toast indicating that the download is complete
        Toast.makeText(appContext,
            activity.getString(R.string.download_complete),
            Toast.LENGTH_SHORT).show()
    }

    private fun setupRecyclerView(mGeoDataList: List<GeoData>) {
        recyclerView!!.adapter =
            ItemRecyclerViewAdapter(mGeoDataList, activity, activity.isTwoPane)
    }

    companion object {
        private const val DOWNLOAD_TIME = 4 // Download time simulation
    }
}